setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/**
+ * Copyright © 2013-2017 Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Setup\Module\Di\Code\Reader;
+
+class FileClassScanner
+{
+    /**
+     * The filename of the file to introspect
+     *
+     * @var string
+     */
+
+    protected $filename;
+
+    /**
+     * The list of classes found in the file.
+     *
+     * @var bool
+     */
+
+    protected $classNames = false;
+
+    /**
+     * @var array
+     */
+
+    protected $tokens;
+
+    /**
+     * Constructor for the file class scanner.  Requires the filename
+     *
+     * @param string $filename
+     */
+
+    public function __construct($filename)
+    {
+        $filename = realpath($filename);
+        if (!file_exists($filename) || !\is_file($filename)) {
+            throw new InvalidFileException(
+                sprintf(
+                    'The file "%s" does not exist or is not a file',
+                    $filename
+                )
+            );
+        }
+        $this->filename = $filename;
+    }
+
+    /**
+     * Retrieves the contents of a file.  Mostly here for Mock injection
+     *
+     * @return string
+     */
+
+    public function getFileContents()
+    {
+        return file_get_contents($this->filename);
+    }
+
+    /**
+     * Extracts the fully qualified class name from a file.  It only searches for the first match and stops looking
+     * as soon as it enters the class definition itself.
+     *
+     * Warnings are suppressed for this method due to a micro-optimization that only really shows up when this logic
+     * is called several millions of times, which can happen quite easily with even moderately sized codebases.
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     * @return array
+     */
+
+    protected function extract()
```

Comment:
> micro optimization is a thing here

This needs to be shown in numbers... For the cyclomatic complexity, not for alternative solution with third-party parser.

>  I would not introduce a full fledged parser until it's proven that the impact is negligible

Ok, no objections :) I was just asking to evaluate possibility of PHP Parser usage here and now, it can be improved later iteratively.